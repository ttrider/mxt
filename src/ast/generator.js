"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const t = __importStar(require("@babel/types"));
const g = __importStar(require("@babel/generator"));
function generateFromAst(ast) {
    if (t.isExpression(ast)) {
        return g.default(ast, {});
    }
    if (t.isStatement(ast)) {
        return g.default(ast, {});
    }
    if (t.isProgram(ast)) {
        return g.default(ast, {});
    }
    if (t.isFile(ast)) {
        return g.default(ast, {});
    }
    if (Array.isArray(ast)) {
        return g.default(t.program(ast), {});
    }
}
exports.generateFromAst = generateFromAst;
//# sourceMappingURL=generator.js.map