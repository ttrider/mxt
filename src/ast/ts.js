"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ts = __importStar(require("typescript"));
function parseInlineExpressions(content) {
    const state = {
        content,
        hasTokens: false,
        externalReferences: []
    };
    const code = "`" + content + "`";
    const result = getDiagnosticsForText(code);
    if (result) {
        state.hasTokens = true;
        state.externalReferences =
            result
                .filter(r => r.code === 2304)
                .map(r => code.substr(r.start === undefined ? 0 : r.start, r.length));
    }
    return state;
}
exports.parseInlineExpressions = parseInlineExpressions;
;
function getDiagnosticsForText(text) {
    const dummyFilePath = "/file.ts";
    const textAst = ts.createSourceFile(dummyFilePath, text, ts.ScriptTarget.Latest);
    const ht = hasTokens(textAst);
    if (ht) {
        const options = {};
        const host = {
            fileExists: filePath => filePath === dummyFilePath,
            directoryExists: dirPath => dirPath === "/",
            getCurrentDirectory: () => "/",
            getDirectories: () => [],
            getCanonicalFileName: fileName => fileName,
            getNewLine: () => "\n",
            getDefaultLibFileName: () => "",
            getSourceFile: filePath => filePath === dummyFilePath ? textAst : undefined,
            readFile: filePath => filePath === dummyFilePath ? text : undefined,
            useCaseSensitiveFileNames: () => true,
            writeFile: () => { }
        };
        const program = ts.createProgram({
            options,
            rootNames: [dummyFilePath],
            host
        });
        return ts.getPreEmitDiagnostics(program);
    }
}
function hasTokens(node) {
    let value = false;
    traverse(node);
    return value;
    function traverse(node) {
        if (ts.isTemplateSpan(node)) {
            value = true;
            return;
        }
        ts.forEachChild(node, traverse);
    }
}
//# sourceMappingURL=ts.js.map