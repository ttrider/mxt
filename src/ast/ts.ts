import * as ts from "typescript";
import { ExpressionInfo } from "..";

export function parseInlineExpressions(content: string) {

    const state: ExpressionInfo = {

        content,
        externalReferences: []
    };

    const { program, source } = getProgram("`" + content + "`");

    const tokens = getTokens(source);

    if (tokens.length > 0) {
        const diags = ts.getPreEmitDiagnostics(program);

        if (diags) {
            state.tokens = tokens;
            state.externalReferences =
                diags
                    .filter(r => r.code === 2304)
                    .map(r => content.substr(r.start === undefined ? 0 : r.start - 1, r.length));
        }

    }

    return state;
}


function getProgram(content: string) {
    const dummyFilePath = "/file.ts";
    const source = ts.createSourceFile(dummyFilePath, content, ts.ScriptTarget.Latest);

    const options: ts.CompilerOptions = {};
    const host: ts.CompilerHost = {
        fileExists: filePath => filePath === dummyFilePath,
        directoryExists: dirPath => dirPath === "/",
        getCurrentDirectory: () => "/",
        getDirectories: () => [],
        getCanonicalFileName: fileName => fileName,
        getNewLine: () => "\n",
        getDefaultLibFileName: () => "",
        getSourceFile: filePath => filePath === dummyFilePath ? source : undefined,
        readFile: filePath => filePath === dummyFilePath ? content : undefined,
        useCaseSensitiveFileNames: () => true,
        writeFile: () => { }
    };
    const program = ts.createProgram({
        options,
        rootNames: [dummyFilePath],
        host
    });

    return { program, source };
}


function getTokens(node: ts.SourceFile) {

    const tokens: Array<{ start: number, end: number, text: string }> = [];
    traverse(node);
    return tokens;
    function traverse(nd: ts.Node) {
        if (ts.isTemplateLiteral(nd)) {

            const tn = nd as any;

            if (tn.templateSpans) {

                for (const ts of tn.templateSpans) {

                    if (ts.expression) {
                        tokens.push({
                            start: ts.start - 1,
                            end: ts.end - 1,
                            text: ts.text
                        });
                    }
                }
            }
            return;
        }
        ts.forEachChild(nd, traverse);
    }
}

export function generateCode(node: ts.Node | ts.Statement[]) {

    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });

    if (Array.isArray(node)) {

        const sf = ts.updateSourceFileNode(
            ts.createSourceFile("./dummy.ts", "", ts.ScriptTarget.Latest),
            node);

        return printer.printNode(ts.EmitHint.Unspecified, sf, sf);
    }

    const sf = (!ts.isSourceFile(node)) ? ts.createSourceFile("./dummy.ts", "", ts.ScriptTarget.Latest) : node;
    return printer.printNode(ts.EmitHint.Unspecified, node, sf);
}