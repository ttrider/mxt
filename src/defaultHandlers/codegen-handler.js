"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("@babel/types");
const t = __importStar(require("@babel/types"));
const builder_1 = require("../ast/builder");
function codegen(context, componentFile) {
    componentFile.addImport("autorun", "mobx");
    // create definitions code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];
            const constTemplateName = `${template.id}$$template`;
            const templateLiteral = builder_1.makeTemplateLiteral(template.elements);
            componentFile.initStatements.add(builder_1.declareVar(constTemplateName)
                .const
                .init(builder_1.makeCall("document.createElement", "template").build())
                .build(), builder_1.makeAssignment(constTemplateName + ".innerHTML", templateLiteral));
        }
    }
    // create init function code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];
            const constTemplateName = `${template.id}$$template`;
            const funcBody = builder_1.statementList().add(builder_1.declareVar("component")
                .const
                .init(builder_1.makeCall("document.importNode", t.identifier(constTemplateName), true).build())
                .build());
            const elements = template.tokens.reduce((p, v) => {
                if (p[v.elementId] === undefined) {
                    p[v.elementId] = [v];
                }
                else {
                    p[v.elementId].push(v);
                }
                return p;
            }, {});
            for (const elementId in elements) {
                if (elements.hasOwnProperty(elementId)) {
                    const tokenSet = elements[elementId];
                    const elementOriginalId = tokenSet[0].elementIdOriginal;
                    const elementName = `${elementId}$$element`;
                    const externalReferences = Object.keys(tokenSet.reduce((e, i) => {
                        for (const er of i.externalReferences) {
                            e[er] = null;
                        }
                        return e;
                    }, {})).map(er => { return { name: er }; });
                    funcBody
                        .add(builder_1.declareVar(elementName)
                        .const
                        .init(builder_1.makeCall("component.content.getElementById", elementId).build())
                        .build())
                        .add(types_1.ifStatement(types_1.unaryExpression("!", types_1.identifier(elementName)), builder_1.makeThrow(`missing element: @${elementId}`)))
                        .add(builder_1.makeAssignment(`${elementName}.id`, elementOriginalId))
                        .add(builder_1.makeCall("autorun", builder_1.declareFunction()
                        .body(builder_1.declareObjectDestruction(...externalReferences).const.init(types_1.identifier("data")).build())
                        .body(...tokenSet.map(token => builder_1.makeCall(elementName + ".setAttribute", token.attributeName, builder_1.makeTemplateLiteral(token.content)).build()))
                        .build()).build());
                }
            }
            funcBody.add(types_1.ifStatement(types_1.identifier("host"), types_1.expressionStatement(types_1.callExpression(types_1.identifier("host.appendChild"), [types_1.identifier("component.content")]))))
                .add(types_1.returnStatement(types_1.identifier("component.content")));
            componentFile.componentStatements.add(builder_1.declareFunction(template.id)
                .param("data")
                .param("host", null, undefined, "Element")
                .body(funcBody)
                .export
                .build());
        }
    }
    return true;
}
exports.codegen = codegen;
//# sourceMappingURL=codegen-handler.js.map