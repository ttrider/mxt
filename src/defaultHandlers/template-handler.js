"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ts_1 = require("../ast/ts");
let idindex = 1;
function parseTemplate(context, componentFile, element) {
    if (element.name.toLowerCase() !== "template") {
        return false;
    }
    if (element.attribs["type"] !== "mxt") {
        return false;
    }
    const template_id = element.attribs.id;
    if (!template_id) {
        componentFile.errors.push(new Error(`template doesn't have an id`));
        return false;
    }
    if (componentFile.templates[template_id]) {
        componentFile.errors.push(new Error(`duplicate template id`));
        return false;
    }
    const template = {
        id: element.attribs.id,
        name: "template",
        attributes: element.attribs,
        elements: [],
        tokens: []
    };
    componentFile.templates[template_id] = template;
    // traverse template
    for (const item of element.children) {
        const element = item;
        template.elements.push(element);
        processItem(template, element);
    }
    function processItem(template, item) {
        switch (item.type.toLowerCase()) {
            case "cdata" /* CDATA */: break;
            case "comment" /* Comment */: break;
            case "directive" /* Directive */: break;
            case "doctype" /* Doctype */: break;
            case "script" /* Script */: break;
            case "style" /* Style */: break;
            case "tag" /* Tag */:
                processTag(template, item);
                break;
            case "text" /* Text */: break;
        }
        //const name = item.name.toLowerCase();
        // if (name === "template") {
        //     // extract sub-template
        //     //pre parse
        //     // parse
        //     // post parse
        //     parseTemplate(context, componentFile, item);
        //     if (item.parent) {
        //         const indexOf = item.parent.children.indexOf(item);
        //         if (indexOf !== -1) {
        //             item.parent.children.splice(indexOf, 1);
        //         }
        //     }
        //     return;
        // }
    }
    function processTag(template, tagItem) {
        const tokenizedAttributes = [];
        for (const attrName in tagItem.attribs) {
            if (tagItem.attribs.hasOwnProperty(attrName)) {
                const attrValue = tagItem.attribs[attrName];
                if (attrValue) {
                    const attrState = ts_1.parseInlineExpressions(attrValue);
                    if (attrState.hasTokens) {
                        if (tagItem.attribs.id$$original === undefined) {
                            tagItem.attribs.id$$original = tagItem.attribs.id === undefined ? "" : tagItem.attribs.id;
                            tagItem.attribs.id = `tagid_${idindex++}`;
                        }
                        attrState.attributeName = attrName;
                        attrState.elementId = tagItem.attribs.id;
                        attrState.elementIdOriginal = tagItem.attribs.id$$original;
                        template.tokens.push(attrState);
                        tokenizedAttributes.push(attrName);
                    }
                }
            }
        }
        for (const tokenizedAttribute of tokenizedAttributes) {
            delete tagItem.attribs[tokenizedAttribute];
        }
        delete tagItem.attribs.id$$original;
        for (const item of tagItem.children) {
            processItem(template, item);
        }
    }
    return true;
}
exports.parseTemplate = parseTemplate;
//# sourceMappingURL=template-handler.js.map