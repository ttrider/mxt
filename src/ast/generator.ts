import * as t from "@babel/types";
import * as g from "@babel/generator";


export function generateFromAst(ast: t.File | t.Program | t.Statement | t.Expression | t.Statement[]) {

    if (t.isExpression(ast)) {
        return g.default(ast, {});
    }
    if (t.isStatement(ast)) {
        return g.default(ast, {});
    }
    if (t.isProgram(ast)) {
        return g.default(ast, {});
    }
    if (t.isFile(ast)) {
        return g.default(ast, {});
    }
    if (Array.isArray(ast)) {
        return g.default(t.program(ast), {});
    }

}