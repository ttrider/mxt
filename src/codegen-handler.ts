import { HandlerContext, AttributeTokenInfo } from ".";
import { returnStatement, unaryExpression, ifStatement, expressionStatement, identifier, callExpression } from "@babel/types";
import * as t from "@babel/types";
import { statementList, declareFunction, declareVar, declareObjectDestruction, makeTemplateLiteral, makeAssignment, makeThrow, makeCall } from "./ast/builder";
import { ComponentFile } from "./component-file";


export function codegen(context: HandlerContext, componentFile: ComponentFile) {

    componentFile.addImport("autorun", "mobx");

    // create definitions code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];

            const constTemplateName = `${template.id}$$template`;

            const templateLiteral = makeTemplateLiteral(template.elements);


            componentFile.initStatements.add(

                declareVar(constTemplateName)
                    .const
                    .init(makeCall("document.createElement", "template").build())
                    .build(),

                makeAssignment(constTemplateName + ".innerHTML", templateLiteral)
            );

        }
    }

    // create init function code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];

            const constTemplateName = `${template.id}$$template`;

            const funcBody = statementList().add(
                declareVar("component")
                    .const
                    .init(makeCall("document.importNode", t.identifier(constTemplateName), true).build())
                    .build()
            );

            const elements = template.tokens.reduce((p, v) => {
                if (p[v.elementId] === undefined) {
                    p[v.elementId] = [v];
                } else {
                    p[v.elementId].push(v);
                }
                return p;
            }, {} as { [elementId: string]: AttributeTokenInfo[] });


            for (const elementId in elements) {
                if (elements.hasOwnProperty(elementId)) {
                    const tokenSet = elements[elementId];
                    const elementOriginalId = tokenSet[0].elementIdOriginal;

                    const elementName = `${elementId}$$element`;

                    const externalReferences = Object.keys(tokenSet.reduce<{ [name: string]: any }>((e, i) => {
                        for (const er of i.externalReferences) {
                            e[er] = null;
                        }
                        return e;
                    }, {})).map(er => { return { name: er }; });





                    funcBody
                        .add(declareVar(elementName)
                            .const
                            .init(makeCall("component.content.getElementById", elementId).build())
                            .build())

                        .add(ifStatement(unaryExpression("!", identifier(elementName)), makeThrow(`missing element: @${elementId}`)))
                        .add(makeAssignment(`${elementName}.id`, elementOriginalId))
                        .add(makeCall("autorun", declareFunction()
                            .body(declareObjectDestruction(...externalReferences).const.init(identifier("data")).build())
                            .body(...tokenSet.map(token => makeCall(elementName + ".setAttribute", token.attributeName, makeTemplateLiteral(token.content)).build()))
                            .build()).build())
                        ;
                }
            }

            funcBody.add(ifStatement(identifier("host"), expressionStatement(callExpression(identifier("host.appendChild"), [identifier("component.content")]))))
                .add(returnStatement(identifier("component.content")));


            componentFile.componentStatements.add(
                declareFunction(template.id)
                    .param("data")
                    .param("host", null, undefined, "Element")
                    .body(funcBody)
                    .export
                    .build());


        }
    }
    return true;
}
