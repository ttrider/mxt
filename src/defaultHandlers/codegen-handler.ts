import { HandlerContext, AttributeTokenInfo } from "..";
import ts from "typescript";
import { statements, declareFunction, declareVar, declareObjectDestruction, makeTemplateLiteral, makeAssignment, makeThrow, makeCall } from "../ts/builder";
import { ComponentFile } from "../component-file";


export function codegen(context: HandlerContext, componentFile: ComponentFile) {

    componentFile.addImport("autorun", "mobx");

    // create definitions code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];

            const constTemplateName = `${template.id}$$template`;

            const templateLiteral = makeTemplateLiteral(template.elements);


            componentFile.initStatements.add(

                declareVar(constTemplateName)
                    .const
                    .init(makeCall("document.createElement", "template")),

                ts.createStatement(makeAssignment(constTemplateName + ".innerHTML", templateLiteral))
            );

        }
    }

    // create init function code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];

            const constTemplateName = `${template.id}$$template`;

            const funcBody = statements().add(
                declareVar("component")
                    .const
                    .init(makeCall("document.importNode", ts.createIdentifier(constTemplateName), true))
            );

            const elements = template.tokens.reduce((p, v) => {
                if (p[v.elementId] === undefined) {
                    p[v.elementId] = [v];
                } else {
                    p[v.elementId].push(v);
                }
                return p;
            }, {} as { [elementId: string]: AttributeTokenInfo[] });


            for (const elementId in elements) {
                if (elements.hasOwnProperty(elementId)) {
                    const tokenSet = elements[elementId];
                    const elementOriginalId = tokenSet[0].elementIdOriginal;

                    const elementName = `${elementId}$$element`;

                    const externalReferences = Object.keys(tokenSet.reduce<{ [name: string]: any }>((e, i) => {
                        for (const er of i.externalReferences) {
                            e[er] = null;
                        }
                        return e;
                    }, {})).map(er => { return { name: er }; });

                    funcBody
                        .add(declareVar(elementName)
                            .const
                            .init(makeCall("component.content.getElementById", elementId).build())
                            .build())

                        .add(ts.createIf(
                            ts.createPrefix(ts.SyntaxKind.ExclamationToken, ts.createIdentifier(elementName)),
                            makeThrow(`missing element: @${elementId}`)
                        ))

                        .add(ts.createStatement(makeAssignment(`${elementName}.id`, elementOriginalId)))
                        .add(makeCall("autorun", declareFunction()
                            .body(declareObjectDestruction(...externalReferences).const.init(ts.createIdentifier("data")))
                            .body(...tokenSet.map(token => makeCall(elementName + ".setAttribute", token.attributeName, makeTemplateLiteral(token.content)).build()))
                            .build()).build())
                        ;
                }
            }

            funcBody.add(ts.createIf(ts.createIdentifier("host"), ts.createStatement(makeCall("host.appendChild", ts.createIdentifier("component.content")).build())))
                .add(ts.createReturn(ts.createIdentifier("component.content")));

            componentFile.componentStatements.add(
                declareFunction(template.id)
                    .param("data")
                    .param("host", null, undefined, "Element")
                    .body(funcBody.build())
                    .export
                    .build());


        }
    }
    return true;
}
