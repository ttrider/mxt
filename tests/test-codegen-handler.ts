import { parseTemplate } from "../src/defaultHandlers/template-handler";
import { codegen } from "../src/defaultHandlers/codegen-handler";
import { setupElementTest } from "./utils";
import { generateFromAst } from "../src/ast/generator";
import { generateCode } from "../src/ast/ts";

describe("token", () => {

  test("single element, attribute, token", () => {

    const input =
      `<template id="t01" type="mxt">
    <div id="old" style="color: \${color}">Hello MXT!</div>
  </template>`;

    const results = setup(input);
    expect(results).toMatchSnapshot();
  });

  test("single element, multiple attributes,tokens", () => {

    const input =
      `<template id="t01" type="mxt">
    <div id="old" style="color: \${color}" title="\${title}">Hello MXT!</div>
  </template>`;

    const results = setup(input);
    expect(results).toMatchSnapshot();
  });

  test("single element, multiple tokens per attribute", () => {

    const input =
      `<template id="t01" type="mxt">
    <div id="old" style="color: \${color}; margin: \${margin}" >Hello MXT!</div>
  </template>`;

    const results = setup(input);
    expect(results).toMatchSnapshot();
  });

  test("multiple elements", () => {

    const input =
      `<template id="t01" type="mxt">
    <div style="color: \${color};">Hello MXT!</div>
    <div title="\${title}">Hello MXT too!</div>
  </template>`;

    const results = setup(input);
    expect(results).toMatchSnapshot();
  })

  function setup(content: string) {
    const { context, component, element } = setupElementTest(content);

    expect(parseTemplate(context, component, element)).toBe(true);
    expect(codegen(context, component)).toBe(true);
    return generateCode(component.getStatements())
  }
});
