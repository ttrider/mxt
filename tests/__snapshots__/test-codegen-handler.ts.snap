// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`token multiple elements 1`] = `
Object {
  "code": "import { autorun } from \\"mobx\\";
const t01$$template = document.createElement(\\"template\\");
t01$$template.innerHTML = \`
    <div id=\\"tagid_4\\">Hello MXT!</div>
    <div id=\\"tagid_5\\">Hello MXT too!</div>
  \`;
export function t01(data: any, host?: null | undefined | Element) {
  const component = document.importNode(t01$$template, true);
  const tagid_4$$element = component.content.getElementById(\\"tagid_4\\");
  if (!tagid_4$$element) throw new Error(\\"missing element: @tagid_4\\");
  tagid_4$$element.id = \\"\\";
  autorun(() => {
    const {
      color
    } = data;
    tagid_4$$element.setAttribute(\\"style\\", \`color: \${color};\`);
  });
  const tagid_5$$element = component.content.getElementById(\\"tagid_5\\");
  if (!tagid_5$$element) throw new Error(\\"missing element: @tagid_5\\");
  tagid_5$$element.id = \\"\\";
  autorun(() => {
    const {
      title
    } = data;
    tagid_5$$element.setAttribute(\\"title\\", \`\${title}\`);
  });
  if (host) host.appendChild(component.content);
  return component.content;
}",
  "map": null,
  "rawMappings": null,
}
`;

exports[`token single element, attribute, token 1`] = `
Object {
  "code": "import { autorun } from \\"mobx\\";
const t01$$template = document.createElement(\\"template\\");
t01$$template.innerHTML = \`
    <div id=\\"tagid_1\\">Hello MXT!</div>
  \`;
export function t01(data: any, host?: null | undefined | Element) {
  const component = document.importNode(t01$$template, true);
  const tagid_1$$element = component.content.getElementById(\\"tagid_1\\");
  if (!tagid_1$$element) throw new Error(\\"missing element: @tagid_1\\");
  tagid_1$$element.id = \\"old\\";
  autorun(() => {
    const {
      color
    } = data;
    tagid_1$$element.setAttribute(\\"style\\", \`color: \${color}\`);
  });
  if (host) host.appendChild(component.content);
  return component.content;
}",
  "map": null,
  "rawMappings": null,
}
`;

exports[`token single element, multiple attributes,tokens 1`] = `
Object {
  "code": "import { autorun } from \\"mobx\\";
const t01$$template = document.createElement(\\"template\\");
t01$$template.innerHTML = \`
    <div id=\\"tagid_2\\">Hello MXT!</div>
  \`;
export function t01(data: any, host?: null | undefined | Element) {
  const component = document.importNode(t01$$template, true);
  const tagid_2$$element = component.content.getElementById(\\"tagid_2\\");
  if (!tagid_2$$element) throw new Error(\\"missing element: @tagid_2\\");
  tagid_2$$element.id = \\"old\\";
  autorun(() => {
    const {
      color,
      title
    } = data;
    tagid_2$$element.setAttribute(\\"style\\", \`color: \${color}\`);
    tagid_2$$element.setAttribute(\\"title\\", \`\${title}\`);
  });
  if (host) host.appendChild(component.content);
  return component.content;
}",
  "map": null,
  "rawMappings": null,
}
`;

exports[`token single element, multiple tokens per attribute 1`] = `
Object {
  "code": "import { autorun } from \\"mobx\\";
const t01$$template = document.createElement(\\"template\\");
t01$$template.innerHTML = \`
    <div id=\\"tagid_3\\">Hello MXT!</div>
  \`;
export function t01(data: any, host?: null | undefined | Element) {
  const component = document.importNode(t01$$template, true);
  const tagid_3$$element = component.content.getElementById(\\"tagid_3\\");
  if (!tagid_3$$element) throw new Error(\\"missing element: @tagid_3\\");
  tagid_3$$element.id = \\"old\\";
  autorun(() => {
    const {
      color,
      margin
    } = data;
    tagid_3$$element.setAttribute(\\"style\\", \`color: \${color}; margin: \${margin}\`);
  });
  if (host) host.appendChild(component.content);
  return component.content;
}",
  "map": null,
  "rawMappings": null,
}
`;
